lib/hanami/action/flash.rb

```diff
@@ -125,94 +133,124 @@ def inspect
       #
       # @since 0.3.0
       # @api private
-      def flash
+      def _data
         @session[SESSION_KEY] || {}
       end
 
-      # The flash registry that holds the data **only for** the current request
+      # Remove the flash entirely from the session if empty.
       #
-      # @return [Hash] the flash for the current request
+      # @return [void]
       #
       # @since 0.3.0
       # @api private
-      def data
-        flash[@request_id] || {}
+      #
+      # @see Hanami::Action::Flash#empty?
+      def remove!
+        if empty?
+          @session.delete(SESSION_KEY)
+          @session.delete(KEPT_KEY)
+        end
       end
 
-      # Expire the stale data from the previous request.
+      # Values from session and kept
       #
-      # @return [void]
+      # @return [Hash]
       #
       # @since 0.3.0
       # @api private
-      def expire_stale!
-        flash.each do |request_id, _|
-          flash.delete(request_id) if delete?(request_id)
-        end
+      def _values
+        _data.merge(kept_data)
       end
 
-      # Remove the flash entirely from the session if empty.
+      # Get the kept request data
       #
-      # @return [void]
+      # @return [Array]
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      #
-      # @see Hanami::Action::Flash#empty?
-      def remove!
-        @session.delete(SESSION_KEY) if empty?
+      def kept
+        @session[KEPT_KEY] || []
       end
 
-      # @since 1.0.0
-      # @api private
+      # Merge current data into KEPT_KEY hash
+      #
+      # @return [Hash] the current value of KEPT_KEY
       #
-      # @see Hanami::Action::Flash#[]
-      def merge!
-        @merged = last_request_flash.merge(data)
+      # @since x.x.x
+      # @api private
+      def keep_data
+        new_kept_data = kept << { count: 0, data: _data }.to_json
+
+        update_kept!(new_kept_data)
       end
 
-      # Values from all the stored requests
+      # Removes from kept data those who how live more than 2 requests
       #
-      # @return [Array]
+      # @return [Hash] the current value of KEPT_KEY
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      def _values
-        flash.select { |session_id, _| !delete?(session_id) }.values
+      def expire_kept!
+        new_kept_data = kept.reject do |kept_data|
+          JSON.parse(kept_data)['count'] >= 2
```

Comment:
@GustavoCaso I think we should be defensive here and don't assume that this `count` is 1. present 2. an integer. If session data is modified (read evil craft), this would crash.

Reply:
@jodosha do you like the idea of:

```ruby
count = JSON.parse(kept_data)['count']
count.to_i >= 2
```

In the case `count` is not present using `to_i` will return `0` and ultimatley the condition will evaluate to `false`